dotnet/src/Connectors/Connectors.AI.Oobabooga/TextCompletion/OoobaboogaTextCompletion.cs

```diff
@@ -0,0 +1,232 @@
+        {
+            var blockingUri = new UriBuilder(this._endpoint)
+            {
+                Port = this._blockingPort,
+                Path = BlockingUriPath
+            };
+
+            var completionRequest = this.CreateOobaboogaRequest(text, requestSettings);
+
+            using var stringContent = new StringContent(
+                JsonSerializer.Serialize(completionRequest),
+                Encoding.UTF8,
+                "application/json");
+
+            using var httpRequestMessage = new HttpRequestMessage()
+            {
+                Method = HttpMethod.Post,
+                RequestUri = blockingUri.Uri,
+                Content = stringContent
+            };
+
+            using var response = await this._httpClient.SendAsync(httpRequestMessage, cancellationToken).ConfigureAwait(false);
+            response.EnsureSuccessStatusCode();
+
+            var body = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
+
+            TextCompletionResponse? completionResponse = JsonSerializer.Deserialize<TextCompletionResponse>(body);
+
+            if (completionResponse is null)
+            {
+                throw new AIException(AIException.ErrorCodes.InvalidResponseContent, "Unexpected response from model")
+                {
+                    Data = { { "ModelResponse", body } },
+                };
+            }
+
+            return completionResponse.Results.Select(completionText => new TextCompletionResult(completionText.Text ?? string.Empty)).ToList();
+        }
+        catch (Exception e) when (e is not AIException && !e.IsCriticalException())
+        {
+            throw new AIException(
+                AIException.ErrorCodes.UnknownError,
+                $"Something went wrong: {e.Message}", e);
+        }
+    }
+
+    /// <inheritdoc/>
+    public void Dispose()
+    {
+        this._httpClient.Dispose();
+        this._webSocket.Dispose();
+        this._httpClientHandler?.Dispose();
+    }
+
+    #region private ================================================================================
+
+    /// <summary>
+    /// Creates an Oobabooga request.
+    /// </summary>
+    /// <param name="text">The text to complete.</param>
+    /// <param name="requestSettings">The request settings.</param>
+    /// <returns>An OobaboogaTextCompletionRequest.</returns>
+    private TextCompletionRequest CreateOobaboogaRequest(string text, CompleteRequestSettings requestSettings)
+    {
+        if (string.IsNullOrWhiteSpace(text))
+        {
+            throw new ArgumentNullException(nameof(text));
+        }
+
+        // Prepare the request using the provided parameters.
+        return new TextCompletionRequest()
+        {
+            Prompt = text,
+            MaxNewTokens = requestSettings.MaxTokens,
+            Temperature = requestSettings.Temperature,
+            TopP = requestSettings.TopP,
+            RepetitionPenalty = 0.5f + requestSettings.PresencePenalty,
```

Comment:
See [#1f54bcc13285571513d0f0cadd9781f80cb3d290](https://github.com/microsoft/semantic-kernel/pull/1357/commits/1f54bcc13285571513d0f0cadd9781f80cb3d290)
As I wrote in my initial PR comment, parameter matching would probably need more attention since the scales don't match.
This one matches a -2:+2 scale to an exponential 0:+inf one. The new formula should be OK except for -2 which will map to 0 and generate an exception on the Oobabooga side, which expects a strictly positive number. 
I thought about adding double.Epsilon by default to avoid that, but then neutral 1 isn't exactly 1 anymore even if very close. I'm not sure which solution is best.
More generally, there seem to be different parameters and different interpretation of those parameters accross the LLM landscape while this isn't propertly standardised. 
Obviously, oobabooga supporting everything tries to accommodate that in a ad-hoc manner, and there isn't probably an ideal answer to those kinds of questions yet. 
It's probably not too important for now though, since semantic-kernel only supports a basic handful of those parameters.